indexing-service/src/main/java/io/druid/indexing/overlord/ForkingTaskRunner.java

```diff
@@ -388,16 +760,191 @@ public void shutdown(final String taskid)
+        );
+    if (taskDirFileList.isEmpty()) {
+      log.info("No task dirs found in [%s]", baseDir);
+      return;
+    }
+    for (File potentialTaskDir : taskDirFileList) {
+      final File[] taskAttemptDirFileArray = potentialTaskDir.listFiles();
+      final Collection<File> taskAttemptDirFileList =
+          taskAttemptDirFileArray == null ?
+          ImmutableList.<File>of() :
+          Collections2.filter(
+              Arrays.asList(taskAttemptDirFileArray),
+              new Predicate<File>()
+              {
+                @Override
+                public boolean apply(File input)
+                {
+                  return input.isDirectory() && input.listFiles(
+                      new FilenameFilter()
+                      {
+                        @Override
+                        public boolean accept(File dir, String name)
+                        {
+                          return TASK_FILE_NAME.equals(name)
+                                 && new File(dir, TASK_FILE_NAME).length()
+                                    > 0;
+                        }
+                      }
+                  ) != null;
+                }
+              }
+          );
+      if (taskAttemptDirFileList.isEmpty()) {
+        log.info("Directory [%s] has no viable task attempts, attempting to cleanup if empty");
+        if (!potentialTaskDir.delete()) {
+          log.warn("Could not clean up [%s]", potentialTaskDir.toString());
+        }
+        continue;
+      }
+      File latestAttemptDir = null;
+      for (File taskAttemptDir : taskAttemptDirFileList) {
+        if (latestAttemptDir == null) {
+          latestAttemptDir = taskAttemptDir;
+        } else {
+          if (latestAttemptDir.lastModified() < taskAttemptDir.lastModified()) {
+            latestAttemptDir = taskAttemptDir;
+          }
+        }
+      }
+      if (latestAttemptDir == null) {
+        log.wtf("I had directories in [%s] but now I don't... memory corruption?", potentialTaskDir.toString());
+        continue;
+      }
+      final File taskFile = new File(latestAttemptDir, TASK_FILE_NAME);
+      final Task task;
+      try {
+        task = jsonMapper.readValue(taskFile, Task.class);
+      }
+      catch (IOException e) {
+        log.makeAlert(e, "Corrupted task file at [%s]", taskFile);
+        log.error(e, "Corrupted task file at [%s]", taskFile);
+        continue;
+      }
+
+      final File statusFile = new File(latestAttemptDir, STATUS_FILE_NAME);
+
+      if (!statusFile.exists()) {
+        // Shouldn't be missing unless there's corruption somehow.
+        log.makeAlert("Status file [%s] is missing ", statusFile);
+        continue;
+      }
+
+      final File portFile = new File(latestAttemptDir, PORT_FILE_NAME);
+      Integer port = null;
+      if (portFile.exists()) {
+        final byte[] buffer = new byte[INT_MAX_STRING_LEN];
+        try (final FileInputStream fileInputStream = new FileInputStream(portFile)) {
```

Comment:
Changed. Realized there was an error mode where the peon is starting up and the ForkingTaskRunner is starting up as well, now it uses file channels with locks.